server/sonar-db-dao/src/main/java/org/sonar/db/permission/GroupPermissionMapper.java

```diff
@@ -51,6 +52,13 @@ void delete(@Param("permission") String permission, @Param("organizationUuid") S
   void selectAllPermissionsByGroupId(@Param("organizationUuid") String organizationUuid,
     @Param("groupId") Integer groupId, ResultHandler resultHandler);
 
+  /**
+   * Lists id of groups with at least one permission on the specified root component but which do not have the specified
+   * permission, <strong>excluding group "AnyOne"</strong> (which implies the returned {@code Sett} can't contain
+   * {@code null}).
+   */
+  Set<Integer> selectGroupIdsWithPermissionOnProjectBut(@Param("projectId") long projectId, @Param("role") String permission);
```

Comment:
yes, it's mandatory, we don't want to retrieve all groups with permission on a given project, it's an unlimited number